torch/_dynamo/variables/torch.py

```diff
@@ -1325,15 +1325,35 @@ def patched_fn(*args, **kwargs):
         # variant torch ops, the original function could come from a user
         # defined `@allow_in_graph` function as well, which doesn't have the
         # same semantics as the torch ops.
-        fake_out_shape = None
-        if "out" in kwargs and isinstance(kwargs["out"], variables.TensorVariable):
-            # Calling fake tensor propagation can mutate the out= tensor in
-            # tx.output.tracked_fakes. tracked_fakes are used to apply
-            # symbolic_shape guards. Mutating them destroys the information
-            # prior to tracing, which is essential for creating right
-            # guards. So save the shape now, and check later if it has
-            # changed. If it has, graph break.
-            fake_out_shape = kwargs["out"].proxy.node.meta["example_value"].shape
+
+        # Calling fake tensor propagation can mutate the out= tensor in
+        # tx.output.tracked_fakes. tracked_fakes are used to apply
+        # symbolic_shape guards. Mutating them destroys the information
+        # prior to tracing, which is essential for creating right
+        # guards. So save the shape now, and check later if it has
+        # changed. If it has, graph break.
+        saved_out_shapes = None
+        out_kwarg_vt = None
+        if "out" in kwargs:
+            out_kwarg_vt = kwargs["out"]
+
+            # e.g., out=(t1, t2, ...)
+            if isinstance(out_kwarg_vt, (TupleVariable, ListVariable)):
+                temp = []
+                all_tensors = True
+                for vt in out_kwarg_vt.items:
+                    if isinstance(vt, variables.TensorVariable):
+                        shape = vt.proxy.node.meta["example_value"].shape
+                        temp.append(shape)
+                    else:
+                        all_tensors = False
+                        break
+                if all_tensors:
+                    saved_out_shapes = temp
```

Comment:
why don't we save the shapes for a list of mixed Tensor and None? (maybe the answer is that we don't have an example for this)

Reply:
> (maybe the answer is that we don't have an example for this)

Yeah, but I just tested and it errors in eager, so to fully make this compatible, I can raise an user error here (the old impl silently ignores the error, like this PR does now).

Reply:
sounds good

Reply:
Actually I just kept it as a list of mixed Tensor and None, because otherwise it could be BC breaking in theory... https://github.com/pytorch/pytorch/blob/e15ea965a1e84029eb900b93f5776767a39fd91e/torch/_dynamo/variables/torch.py#L1324-L1327